sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/PartitionBasedLoadBalancer.java

```diff
@@ -0,0 +1,333 @@
+            .map(this::loadBalance)
+            // if there was an error, log warning and TODO: call user provided error handler
+            .doOnError(ex -> logger.warning("Load balancing for event processor failed - {}", ex.getMessage()))
+            .subscribe();
+    }
+
+    /*
+     * This method works with the given partition ownership details and Event Hub partitions to evaluate whether the
+     * current Event Processor should take on the responsibility of processing more partitions.
+     */
+    private Mono<Void> loadBalance(final Tuple2<Map<String, PartitionOwnership>, List<String>> tuple) {
+
+        Map<String, PartitionOwnership> partitionOwnershipMap = tuple.getT1();
+        List<String> partitionIds = tuple.getT2();
+
+        if (ImplUtils.isNullOrEmpty(partitionIds)) {
+            // ideally, should never happen.
+            return Mono.error(new IllegalStateException("There are no partitions in Event Hub " + this.eventHubName));
+        }
+
+        if (!isValid(partitionOwnershipMap)) {
+            // User data is corrupt.
+            return Mono.error(new IllegalStateException("Invalid partitionOwnership data from PartitionManager"));
+        }
+
+        /*
+         * Remove all partitions ownerships that have not be modified for a long time. This means that the previous
+         * event processor that owned the partition is probably down and the partition is now eligible to be
+         * claimed by other event processors.
+         */
+        Map<String, PartitionOwnership> activePartitionOwnershipMap = removeInactivePartitionOwnerships(
+            partitionOwnershipMap);
+
+        if (ImplUtils.isNullOrEmpty(activePartitionOwnershipMap)) {
+            /*
+             * If the active partition ownership map is empty, this is the first time an event processor is
+             * running or all Event Processors are down for this Event Hub, consumer group combination. All
+             * partitions in this Event Hub are available to claim. Choose a random partition to claim ownership.
+             */
+            claimOwnership(partitionOwnershipMap, partitionIds.get(RANDOM.nextInt(partitionIds.size())));
+            return Mono.empty();
+        }
+
+        /*
+         * Create a map of owner id and a list of partitions it owns
+         */
+        Map<String, List<PartitionOwnership>> ownerPartitionMap = activePartitionOwnershipMap.values()
+            .stream()
+            .collect(
+                Collectors.groupingBy(PartitionOwnership::ownerId, mapping(Function.identity(), toList())));
+
+        // add the current event processor to the map if it doesn't exist
+        ownerPartitionMap.putIfAbsent(this.ownerId, new ArrayList<>());
+
+        /*
+         * Find the minimum number of partitions every event processor should own when the load is
+         * evenly distributed.
+         */
+        int minPartitionsPerEventProcessor = partitionIds.size() / ownerPartitionMap.size();
```

Comment:
nit: This may be a lack of understanding of language idioms, but you're making several calls to the `size()` methods in this logic.   Maybe it is worth considering reading them once and caching in a local variable for efficiency?